designs/s3/design-vscode-s3.md

```diff
@@ -0,0 +1,262 @@
+    - If user mouse-hovers the folder, show a tooltip like `Filter: "foo bar"`.
+  - Mouse-hover on a bucket/folder reveals [inline](https://code.visualstudio.com/api/extension-guides/tree-view#view-actions)
+    `Filter...` button (so user does not need to dig into the context-menu).
+  - Context-menu includes `Filter...` menu item
+
+### S3 root node
+
+The S3 root node is the "meta node" which lists available buckets in the
+current region.
+
+- Toolkit toolbar:
+  <br><img src="design-vscode-s3-toolbar.png" width="200"/>
+  - (P0) When any S3 node (bucket/folder/object) is selected, Toolkit toolbar
+    enables an `Upload...` button.
+    - If a non-folder object is selected, `Upload...` targets the object's parent folder.
+- S3 root node context-menu:
+  - (P0) `New bucket`
+    - Invokes the `AWS: New bucket` VSCode command which prompts for Bucket name.
+    - Public access is blocked by default. To enable public access, user must
+      use the AWS web console.
+
+### S3 buckets
+
+Buckets are listed as children of the S3 root node.
+
+- Bucket node context-menu:
+  - (P0) `Copy ARN`
+  - (P0) `Copy name`
+  - (P0) `New folder...`
+  - (P0) `Upload...`
+    - Shows VSCode-standard "Open file..." dialog.
+    - Shows a progress dialog using VSCode conventions.
+    - When upload succeeds or fails, shows a toaster summarizing which uploads succeeded and failed.
+    - (P1) Allows multiple files and folders to be uploaded. Creates S3 folders
+      as needed to match the hierarchy of uploaded folders.
+    - (P2) [Auto-detects Content Type ](https://github.com/aws/aws-toolkit-visual-studio/issues/64)
+  - (P0) `Upload to parent folder...`
+    - Works like `Upload...`, except targeting the parent of the current item.
+      This allows the user to avoid needing to navigate to the parent folder.
+  - _Destructive operations (separator):_
+    - (P1) `Delete bucket`
+      - "Super confirm": similar to AWS web console, show a prompt which
+        requires typing the word "Delete".
+
+### S3 objects
+
+S3 objects and folders are listed as children of their respective parent bucket/folder.
+
+- (P0) Clicking an object in the tree selects it, but does not perform any action.
+- (P1) Double-clicking an item downloads it and opens it in readonly mode (see `Edit` action).
+  - If item is 4MB+, prompt user to ask if download should begin.
+- (P2) User input activates standard VSCode local, as-you-type filtering.
+- S3 object node (files and folders) context-menu:
+  - (P0) `Copy ARN`
+  - (P0) `Copy URL`
+  - (P0) `Copy name`
+  - (P0) `Download`
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default location (typically `~/Downloads`).
+    - Present download status as a toaster message containing this info:
+      - Path to the download location.
+        - On completion, the path becomes a linkbutton, which onclick opens the file in VSCode.
+      - `Choose location...` linkbutton.
```

Comment:
The Choose location is a confusing experience. It applies to configurations for future use, and has no effect on the download in progress. Rather than having a button, what if the notice stated that this is configurable through the toolkit configurations.

Reply:
What if you had a pop up on first download that let you choose a default location? Suggestions can be provided (current workspace folder, downloads, somewhere else). That way, the location is clear to the user and we don't block future downloads.

Reply:
>  It applies to configurations for future use, and has no effect on the download in progress.

Modified to also affect the current in-progress download.

Reply:
Why is it important to be able to change the location while a download is in-progress? This has edge-case impacts with respect to when a download completes.

Reply:
? Downloads typically go a temporary location and must be renamed on completion, regardless of this feature. Thus changing the "download location" just changes the final rename step.

Reply:
Notifications have to be discarded by the user. What happens when the user comes to a stale notification and presses the button, but they have already gone and move/deleted/consumed the file? Handling the rename seems to have race condition implications with the timing of changing the location and when a download completes.

The "Download to..." flow seems to provide all of the necessary utility without additional complexity. Suggest to make this part P1 instead of P0, and allow for more planning around this one.

Reply:
Marked as P1. Agree this is a tricky interaction, I think the description here is enough to inform the direction and the details can be worked out at impl-time.